core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlPrepareAsyncProcessor.java

```diff
@@ -18,31 +18,112 @@
+
   protected final Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache;
 
   public CqlPrepareAsyncProcessor() {
-    this(CacheBuilder.newBuilder().weakValues().build());
+    this(Optional.empty());
+  }
+
+  public CqlPrepareAsyncProcessor(@NonNull Optional<? extends DefaultDriverContext> context) {
+    this(CacheBuilder.newBuilder().weakValues().build(), context);
   }
 
   protected CqlPrepareAsyncProcessor(
-      Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache) {
+      Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache,
+      Optional<? extends DefaultDriverContext> context) {
+
     this.cache = cache;
+    context.ifPresent(
+        (ctx) -> {
+          LOG.info("Adding handler to invalidate cached prepared statements on type changes");
+          EventExecutor adminExecutor = ctx.getNettyOptions().adminEventExecutorGroup().next();
+          ctx.getEventBus()
+              .register(
+                  TypeChangeEvent.class, RunOrSchedule.on(adminExecutor, this::onTypeChanged));
+        });
+  }
+
+  private static boolean typeMatches(UserDefinedType oldType, DataType typeToCheck) {
+
+    switch (typeToCheck.getProtocolCode()) {
+      case ProtocolConstants.DataType.UDT:
+        UserDefinedType udtType = (UserDefinedType) typeToCheck;
+        return udtType.equals(oldType)
+            ? true
+            : Iterables.any(udtType.getFieldTypes(), (testType) -> typeMatches(oldType, testType));
+      case ProtocolConstants.DataType.LIST:
+        ListType listType = (ListType) typeToCheck;
+        return typeMatches(oldType, listType.getElementType());
+      case ProtocolConstants.DataType.SET:
+        SetType setType = (SetType) typeToCheck;
+        return typeMatches(oldType, setType.getElementType());
+      case ProtocolConstants.DataType.MAP:
+        MapType mapType = (MapType) typeToCheck;
+        return typeMatches(oldType, mapType.getKeyType())
+            || typeMatches(oldType, mapType.getValueType());
+      case ProtocolConstants.DataType.TUPLE:
+        TupleType tupleType = (TupleType) typeToCheck;
+        return Iterables.any(
+            tupleType.getComponentTypes(), (testType) -> typeMatches(oldType, testType));
+      default:
+        return false;
+    }
+  }
+
+  private void onTypeChanged(TypeChangeEvent event) {
+    for (Map.Entry<PrepareRequest, CompletableFuture<PreparedStatement>> entry :
+        this.cache.asMap().entrySet()) {
+
+      try {
+        PreparedStatement stmt = entry.getValue().get();
+        if (Iterables.any(
+                stmt.getResultSetDefinitions(), (def) -> typeMatches(event.oldType, def.getType()))
+            || Iterables.any(
+                stmt.getVariableDefinitions(),
+                (def) -> typeMatches(event.oldType, def.getType()))) {
+
+          this.cache.invalidate(entry.getKey());
+          this.cache.cleanUp();
```

Comment:
Make sure we see the invalidation as quickly as possible.  It's a heavier-weight operation but since changes to this cache should be relatively infrequent it seems like a reasonable compromise.